### PR TITLE
Add support for doctrine collections

### PR DESCRIPTION
**Pull request by Spea**
We have several fields in our models that use the `ArrayCollection` type (see [Type annotation/attribute documentation](http://jmsyst.com/libs/serializer/master/reference/annotations#type)). With these changes it is possible to to handle those types differently in the `liip-serializer` packages.  

I have already prepared the changes for the `liip-serializer`, but before I create the PR, I wanted to make sure this is the correct approach.

Side note: At first I created a new `PropertyTypeArrayCollection` class that handles this case, but this got overly complicated to handle so I went with adding a new property `isCollection` to the existing `PropertyArray` class

### CODE REVIEW

File: src/ModelParser/JMSParser.php [hunk 1]
```diff
@@ -70,6 +71,10 @@ abstract class BaseJMSParser implements ModelParserInterface
 
     public function __construct(Reader $annotationsReader)
     {
+        if (PHP_VERSION_ID >= 80000) {
+            $annotationsReader = new AttributeReader($annotationsReader);
```
Comment (dbu) [inline src/ModelParser/JMSParser.php:75]: this is unrelated the the collection change. does it currently add anything, or would we need to do further steps to actually process annotations?

i really like handling attributes, but we should probably do that in a separate MR and add some test cases for the JMSParser to make sure it works as expected.

File: src/Metadata/PropertyTypeArray.php [hunk 2]
```diff
@@ -30,6 +36,9 @@ public function __toString(): string
             return 'array';
         }
         $array = $this->isHashmap() ? '[string]' : '[]';
+        if ($this->isCollection) {
+            $array .= '|Collection';
```
Comment (dbu) [inline src/Metadata/PropertyTypeArray.php:40]: i think that when we don't know the type, we should still return `array|Collection` a few lines above

File: src/TypeParser/JMSTypeParser.php [hunk 3]
```diff
@@ -65,12 +66,13 @@ private function parseType(array $typeInfo, bool $isSubType = false): PropertyTy
             return new PropertyTypeClass($typeInfo['name'], $nullable);
         }
 
-        if (self::TYPE_ARRAY === $typeInfo['name']) {
+        $isCollection = self::TYPE_ARRAY_COLLECTION === $typeInfo['name'];
```
Comment (dbu) [inline src/TypeParser/JMSTypeParser.php:69]: couldn't people also be using `Doctrine\Common\Collections\ArrayCollection` in their annotation?

could we respect type imports and do an is_subclass_of check?

File: src/TypeParser/JMSTypeParser.php [hunk 4]
```diff
@@ -17,6 +17,7 @@
 final class JMSTypeParser
 {
     private const TYPE_ARRAY = 'array';
+    private const TYPE_ARRAY_COLLECTION = 'ArrayCollection';
```
Comment (dbu) [inline src/TypeParser/JMSTypeParser.php:20]: should we detect any `ArrayAccess` as array? 

the problem however is that then we would need a hard dependency on doctrine collections to be sure we have something we can instantiate - we can't instantiate generically an unknown type of array access as we don't know its constructor signature.

did you look into this? or should we just leave it at the doctrine class for now? it should be simple to expand later on if somebody has the need.

File: src/TypeParser/PhpTypeParser.php [hunk 5]
```diff
@@ -20,12 +20,18 @@ final class PhpTypeParser
     private const TYPE_MIXED = 'mixed';
     private const TYPE_RESOURCE = 'resource';
     private const TYPE_ARRAY = 'array';
+    private const TYPE_ARRAY_COLLECTION = 'ArrayCollection';
+    private const TYPE_COLLECTION = 'Collection';
```
Comment (dbu) [inline src/TypeParser/PhpTypeParser.php:24]: why do you support collection here but not in the jms parser?

should we do Collection::class everywhere and use is_subclass_of to handle inheritance correctly? (also this could clash with some other class called Collection if we don't use the namespace)

as all collections have ArrayAccess, i think it is correct to support generic Collection. afaik doctrine itself works like this, where you put an ArrayCollection on an entity field, and when you load from the database you get some other type of collection that dynamically loads the referenced entities on demand.

File: tests/ModelParser/JMSParserTest.php [hunk 6]
```diff
@@ -166,6 +166,30 @@ public function providePropertyType(): iterable
...
+        yield [
+            new class() {
+                /**
+                 * @JMS\Type("ArrayCollection<string, int>")
+                 */
+                private $property;
+            },
+            PropertyTypeArray::class,
+            true,
+            'int[string]|Collection|null',
```
Comment (dbu) [inline tests/ModelParser/JMSParserTest.php:191]: the correct thing here would be `Collection<string>` i think.

the string representation is meant for debugging, we should have done `int<string>` i think, but we should not change that now to avoid a BC break.

**Review by dbu — COMMENTED** [review]
very cool, thanks a lot for this pull request! i like the approach with the `isCollection` flag.

i commented a few details, mainly i think we need to take namespace into account as an annotation using the fully qualified namespace should also work, and we should not get tripped up by `Collection` from another namespace.

can you please look into that? 

i'd say lets wrap this up and only address using it in the serializer once we merged this MR. should be relatively easy i think.

File: src/TypeParser/JMSTypeParser.php (hunk 4 above)
Comment (dbu) [reply src/TypeParser/JMSTypeParser.php:20]: thinking of the serializer as well: lets drop this idea - doctrine Collection is good enough. generic array access does not tell how to serialize it either, so its not really feasible.

File: src/ModelParser/JMSParser.php (hunk 1 above)
Comment (Spea) [reply src/ModelParser/JMSParser.php:75]: You are right, I'll move this to a separate PR to keep things separate and also add some tests for this.

File: src/TypeParser/JMSTypeParser.php (hunk 3 above)
Comment (Spea) [reply src/TypeParser/JMSTypeParser.php]: Fixed and adjusted via 50d0172cea019bfc227eacd8aacebc4911a31d61 - I had to still support `ArrayCollection` as this is a keyword that is supported in the in the [ArrayCollectionHandler from JMS](https://github.com/schmittjoh/serializer/blob/master/src/Handler/ArrayCollectionHandler.php#L23)

File: src/TypeParser/PhpTypeParser.php (hunk 5 above)
Comment (Spea) [reply src/TypeParser/PhpTypeParser.php]: I know check for the actual resolved class via d4801f965e1aa72c2b70d4d8c9b943acb3fb1e49 - This makes sure that only collection classes from doctrine are treated as such.

File: src/ModelParser/JMSParser.php (hunk 1 above)
Comment (Spea) [reply src/ModelParser/JMSParser.php]: Implemented via #30

File: src/TypeParser/PhpTypeParser.php [hunk 7]
```diff
@@ -56,14 +57,25 @@ public function parseAnnotationType(string $rawType, \ReflectionClass $declaring
             }
         }
 
-        if (0 === \count($types)) {
+        $isCollection = false;
+        $filteredTypes = [];
+        foreach ($types as $type) {
+            $resolvedClass = $this->resolveClass($type, $declaringClass);
+            if (Collection::class === $resolvedClass || is_subclass_of($resolvedClass, Collection::class)) {
```
Comment (dbu) [inline src/TypeParser/PhpTypeParser.php:64]: we could use `is_a` here, with 3rd parameter `true` to avoid needing 2 conditions: https://3v4l.org/1v6N1

File: tests/ModelParser/JMSParserTest.php [hunk 8]
```diff
@@ -167,6 +167,42 @@ public function providePropertyType(): iterable
...
+        yield [
+            new class() {
+                /**
+                 * @JMS\Type("ArrayCollection<string>")
+                 */
+                private $property;
+            },
+            PropertyTypeArray::class,
+            true,
+            'string[]|\Doctrine\Common\Collections\Collection|null',
```
Comment (dbu) [inline tests/ModelParser/JMSParserTest.php:180]: could we make it output `'string[]|\Doctrine\Common\Collections\Collection<string>|null'` ?

File: tests/ModelParser/JMSParserTest.php [hunk 9]
```diff
@@ -167,6 +167,42 @@ public function providePropertyType(): iterable
...
+        yield [
+            new class() {
+                /**
+                 * @JMS\Type("ArrayCollection<string, int>")
+                 */
+                private $property;
+            },
+            PropertyTypeArray::class,
+            true,
+            'int[string]|\Doctrine\Common\Collections\Collection|null',
```
Comment (dbu) [inline tests/ModelParser/JMSParserTest.php:204]: could this specify key and value, with `...\Collection<int, string>...`?

File: tests/ModelParser/Model/WithImports.php [hunk 10]
```diff
@@ -23,4 +24,9 @@ class WithImports
      * @var Nested
      */
     private $aliasSameNamespace;
+
+    /**
+     * @var Nested[]|Collection
```
Comment (dbu) [inline tests/ModelParser/Model/WithImports.php:29]:
```suggestion
     * @var Nested[]|Collection<Nested>
```

Comment (dbu) [reply tests/ModelParser/Model/WithImports.php:29]: could we test if just `Collection<Nested>` also works correctly, detecting the Nested as element type?

**Comment by dbu** [comment]
thanks, this is shaping up nicely!

File: src/TypeParser/PhpTypeParser.php (hunk 7 above)
Comment (Spea) [reply src/TypeParser/PhpTypeParser.php]: I somehow missed that there is a third parameter. Fixed via 0af690fe5d8e870c1fb7158414f78df3ad8e46ba

File: tests/ModelParser/JMSParserTest.php (hunk 8 above)
Comment (Spea) [reply tests/ModelParser/JMSParserTest.php]: I adjusted the `__toString()` method via a95fd9aec4fb795ac183a9760b8464324ac4e400. But keep in mind that this is only the representation of the property type. Using something like `Collection<string>` in a docblock would not work, as the `PhpTypeParser` does not support this kind of syntax.

File: tests/ModelParser/JMSParserTest.php (hunk 9 above)
Comment (Spea) [reply tests/ModelParser/JMSParserTest.php]: I adjusted the `__toString()` method via a95fd9aec4fb795ac183a9760b8464324ac4e400

File: tests/ModelParser/Model/WithImports.php (hunk 10 above)
Comment (Spea) [reply tests/ModelParser/Model/WithImports.php]: Adjusted via a95fd9aec4fb795ac183a9760b8464324ac4e400

File: src/TypeParser/JMSTypeParser.php [hunk 11]
```diff
@@ -91,4 +94,13 @@ private function parseType(array $typeInfo, bool $isSubType = false): PropertyTy
...
         throw new InvalidTypeException(sprintf('Unknown JMS property found (%s)', var_export($typeInfo, true)));
     }
+
+    private function isCollection(string $name): bool
+    {
+        if (self::TYPE_ARRAY_COLLECTION === $name || Collection::class === $name) {
+            return true;
+        }
+
+        return is_subclass_of($name, Collection::class);
```
Comment (dbu) [inline src/TypeParser/JMSTypeParser.php:104]:
```suggestion
        return is_a($name, Collection::class, true);
```

Comment (dbu) [reply src/TypeParser/JMSTypeParser.php:104]: i'd merge the 2 conditions and do `return self::TYPE_ARRAY_COLLECTION === $name || is_a($name, Collection:class, true);`

Comment (Spea) [reply src/TypeParser/JMSTypeParser.php]: Fixed via 74742d61d256a41dea7dfd43fade461a478bfbf0

**Comment by Spea** [comment]
> can you please add a note in the changelog as well to explain that this new feature has been introduced?

Done, thank you for your extended reviews, highly appreciated :)

File: CHANGELOG.md [hunk 12]
```diff
@@ -3,6 +3,8 @@
 # 0.6.0 (unreleased)
 
 * When running with PHP 8, process attributes in addition to the phpdoc annotations.
+* Support doctrine collections
+* Support identical naming strategy
```
Comment (dbu) [inline CHANGELOG.md:7]:
```suggestion
* Support `identical` property naming strategy
```

**Review by dbu — APPROVED** [review]
thanks a lot!

i will merge it like this and tweak the changelog message (i thought i can accept my suggestion but this seems no longer possible in github)

shall i tag a release immediately or do you first want to play with this in the serializer to make sure it works correctly?

File: CHANGELOG.md (hunk 12 above)
Comment (Spea) [reply CHANGELOG.md]: Thank you - I amended it to the last commit

**Comment by Spea** [comment]
> shall i tag a release immediately or do you first want to play with this in the serializer to make sure it works correctly?

Let me create the PR in the liip-serializer repository first, maybe we have to adjust the parser a bit. Once you are happy with the changes there, you can tag a new release.

**Comment by dbu** [comment]
perfect! thanks a lot for your contributions!
